waspc/data/Generator/templates/sdk/wasp/server/operations/wrappers.ts

```diff
@@ -0,0 +1,233 @@
+ * @returns The server-side API for the provided authenticated operation.
+ */
+export function createAuthenticatedOperation<
+  OperationDefinition extends GenericOperationDefinition
+>(
+  userOperation: OperationDefinition,
+  entities: EntityMapFor<OperationDefinition>
+): AuthenticatedOperationFor<OperationDefinition> {
+  async function operation(...args: AuthenticatedOperationArgsFor<OperationDefinition>) {
+    // To understand this function and its limitations, read
+    // https://github.com/wasp-lang/wasp/issues/2050
+    if (args.length < 1) {
+      // No arguments sent -> no user and no payload specified -> there's no way this was called correctly.
+      throw new Error('Invalid number of arguments')
+    } else if (includesPayload(args)) {
+      // Two arguments sent -> the first argument is the payload, the second is the context.
+      const [payload, context] = args
+      return userOperation(payload, {
+        ...context,
+        entities,
+      })
+    } else {
+      // One argument sent -> the first and only argument is the user.
+      const [context] = args
+      return userOperation(undefined as Parameters<OperationDefinition>[0], {
+        ...context,
+        entities,
+      })
+    }
+  }
+
+  return operation as AuthenticatedOperationFor<OperationDefinition>
+}
+
+/**
+ * Returns a boolean carrying compiler type information about whether the
+ * provided arguments array (of an authenticated operation) includes a payload.
+ *
+ * @template Input The type of the payload the operation expects.
+ * @param args The arguments array for an authenticated operation.
+ * @returns true if the arguments array includes a payload, false otherwise.
+ */
+function includesPayload<Input>(
+  args: [AuthenticatedOperationContext] | [Input, AuthenticatedOperationContext]
+): args is [Input, AuthenticatedOperationContext] {
+  return args.length === 2
+}
+
+/**
+ * Constructs the type of the arguments array for an authenticated operation
+ * based on the type of its definition.
+ */
+type AuthenticatedOperationArgsFor<Op extends GenericOperationDefinition> =
+  Parameters<AuthenticatedOperationFor<Op>>
+
+/**
+ * Constructs the type for an authenticated operation's server-side API.
+ *
+ * @template Input The type of the payload the operation expects (must be `void` if the operation doesn't expect a payload).
+ * @template Output The type of the operation's return value.
+ */
+type AuthenticatedOperation<Input, Output> = Operation<Input, Output, true>
+
+{=/ isAuthEnabled =}
+/**
+ * Constructs the type for an unauthenticated operation's server-side API.
+ *
+ * @template Input The type of the payload the operation expects (must be `void` if the operation doesn't expect a payload).
+ * @template Output The type of the operation's return value.
+ */
+type UnauthenticatedOperation<Input, Output> = Operation<Input, Output, false>
+
+// todo(filip): Should i define this independently of AuthenticatedOperationDefinition?
```

Comment:
Yup, I thought that defining a generic operation via more specific authenticated operation is weird. And I still do.

Defining public API type via internal stuff -> why not?

Reply:
> Defining public API type via internal stuff -> why not?

Because the API can change unexpectedly when we change an internal definition, and the public API should arguably be its own source of truth.

This was one of the core reasons decisions behind structural typing in TS and Go: _define the structure you expect instead of importing names_.

It all depends on where it makes sense to have the source of truth for something. In this case, I was on the edge, but if you don't mind it, I'm keeping it as is.

> Yup, I thought that defining a generic operation via more specific authenticated operation is weird. And I still do.

Yes, it's a little unintuitive, but correct in terms of subtyping and the type system.

There's a comment above it that explains what's going on. You possibly haven't seen it since you reviewed an old version of the code (that's what [this](https://github.com/wasp-lang/wasp/pull/2044#discussion_r1613744651) is about).

So, If you haven't seen it, please take a look at the comment.

If you still don't like it, I'll split the definition into `GenericUnauthenticatedOperation` and `GenericAuthenticatedOperation`.